polynote-server/src/main/scala/polynote/server/SharedNotebook.scala

```diff
@@ -0,0 +1,285 @@
+  * and the highest local version which has been acknowledged by the server (either by sending an edit based on it,
+  * or through an acknowledgement message). It has to keep its local edit history between that known-acknowledged
+  * version and its current version, to allow rebasing new foreign edits.
+  *
+  * It sounds a lot more complicated than it really is.
+  */
+trait SharedNotebook[F[_]] {
+
+  def path: String
+
+  /**
+    * Open a reference to this shared notebook.
+    *
+    * @param name A string identifying who is opening the notebook (i.e. their username or email)
+    * @param oq   A queue to which messages to the subscriber (such as notifications of foreign updates) will be submitted
+    * @return A [[NotebookRef]] which the caller can use to send updates to the shared notebook
+    */
+  def open(name: String, oq: Enqueue[F, Message]): F[NotebookRef[F]]
+
+  def versions: Stream[F, (Int, Notebook)]
+}
+
+
+class IOSharedNotebook(
+  val path: String,
+  ref: SignallingRef[IO, (GlobalVersion, Notebook)],            // the Int is the global version, which can wrap around back to zero if necessary
+  kernelRef: Ref[IO, Option[Kernel[IO]]],
+  updates: Queue[IO, Option[(SubscriberId, NotebookUpdate, Deferred[IO, GlobalVersion])]],   // the canonical series of edits
+  updatesTopic: Topic[IO, Option[(GlobalVersion, SubscriberId, NotebookUpdate)]],  // a subscribe-able channel for watching updates,
+  kernelFactory: KernelFactory[IO],
+  outputMessages: Topic[IO, Message],
+  kernelLock: Semaphore[IO]
+)(implicit contextShift: ContextShift[IO]) extends SharedNotebook[IO] {
+
+  private val updateBuffer = new VersionBuffer[NotebookUpdate]
+
+  private val statusUpdates = Publish.PublishTopic(outputMessages).contramap[KernelStatusUpdate](KernelStatus(ShortString(path), _))
+
+  // listen to the stream of updates and apply them in order, each one incrementing the global version
+  updates.dequeue.unNoneTerminate.zipWithIndex.evalMap {
+    case ((subscriberId, update, versionPromise), version) =>
+      val newGlobalVersion = (version % Int.MaxValue).toInt
+      updateBuffer.add(newGlobalVersion, update)
+      applyUpdate(newGlobalVersion, subscriberId, update, versionPromise)
+  }.compile.drain.unsafeRunAsyncAndForget()
+
+  private def ensureKernel(): IO[Kernel[IO]] = kernelLock.acquire.bracket { _ =>
+    kernelRef.get.flatMap {
+      case None => kernelFactory.launchKernel(() => ref.get.map(_._2), statusUpdates).flatMap {
+        kernel => kernelRef.set(Some(kernel)).as(kernel)
+      }
+      case Some(kernel) => IO.pure(kernel)
+    }
+  }(_ => kernelLock.release)
+
+  // apply a versioned update from the queue, completing its version promise and updating the info about which
+  // versions the originating subscriber knows about
+  private def applyUpdate(newGlobalVersion: Int, subscriberId: Int, update: NotebookUpdate, versionPromise: Deferred[IO, GlobalVersion]) = ref.get.flatMap {
```

Comment:
Use `GlobalVersion` and `SubscriberId` here?

Reply:
Yeah, could do – not sure it's necessary since those are already labeled with argument names. The type aliases were mainly because `(Int, Int, Foo)` tells you nothing about what the `Int`s are for. Maybe shoulda just made a bunch of new wrapper case classes instead... ¯\\\_(ツ)\_/¯

Reply:
Yeah, it just stood out for me because it wasn't consistent with the rest of the code - e.g., we have `private def submitUpdate(subscriberId: SubscriberId ...)` below.